src/hdrhtml/pfsouthdrhtml.1
.TH "pfsouthdrhtml" 1
.SH NAME
pfsouthdrhtml \- Create a web page with an HDR viewer
.SH SYNOPSIS

\fBpfsouthdrhtml\fR [<page_name>] [\fB--quality\fR <1-5>] [\fB--image-dir\fR <directory_name>] [\fB--page-template\fR <template_file>] [\fB--image-template\fR <template_file>] [\fB--object-output\fR <file_name.js>] [\fB--html-output\fR <file_name.html>]

.SH DESCRIPTION
The command creates in the current directory an HTML web page
containing multi-exposure HDR viewer. The multi-exposure viewer
displays a portion of the available dynamic range with minimum
contrast distortions and provides a slider control to move the dynamic
range window towards brighter or darker tones. The interface is very
similar to \fIpfsview\fR, which is a pfstools application for displaying
HDR images. The web page employs only JavaScript and CSS opacity
property and does not require Java applets or the Flash plugin. Note
that because this techniques encodes 20-60 exposures using only few
images, the displayed exposures may not be identical to the exposures
that are shown in pfsview. For examples and more information, visit
.PP
\fIhttp://pfstools.sourceforge.net/hdrhtml/\fR.
.PP
\fI<page_name>\fR specifies the file name, of
the web page to be generated. If \fI<page_name>\fR is missing, the
file name of the first image with .html extension will be used.
.PP
The command can take as input several images and put them all on the
same web page. For each image, its file name (from the FILE_NAME tag
in the pfsstrem) without extension and a leading path will be
used as a name for all JavaScript variables corresponding to that
image. If the filename contains illegal characters (such as
space, '-', '[', etc), these will be converted to '_'.
.TP
\fB--quality\fR <1-5>, \fB-q\fR <1-5>
Quality of the interpolated exposures, from the worst (1) to the best
(5). The default is 2, which is sufficient for most
applications. Higher quality will introduce less distortions in the
brightest and the darkest tones, but will also generate more
images. More images means that there is more data that needs to be
transferred to the web-browser, making HDR viewer less responsive.
.TP
\fB--image-dir\fR <directory_name>, \fB-d\fR <directory_name>
Specify where to store the resulting image files. Links to images in
HTML will be updated accordingly. This must be a relative path and the
directory must exist.  Useful to avoid clutter in the current
directory.
.TP
\fB--page-template\fR <template_file>, \fB-p\fR <directory_name>, \fB--image-template\fR <template_file>, \fB-i\fR <template_file>
Replaces the template files used to generate an HTML web page. The
template files contain all HTML and JaveScript code with special
keywords (@keyword@) that are replaced with image specific data, such
as width, height, image base name, etc. The default template files can
be found in \fIINSTALL_DIR/share/pfstools/hdrhtml_default_templ/hdrhtml_*_templ.html\fR. There is an alternative template bundled with pfstools in the hdrhtml_hdrlabs_templ directory, which contains many improvements and looks much better but requires additional asset files.  The example at the end of this manual shows how to use alternative template. More details on how to design own templates can be found in \fBTEMPLATE FILE FORMAT\fR below.
.TP
\fB--object-output\fR <file_name.js>, \fB-o\fR <file_name.js>
Store JavaScript objects (\fIhdr_<base_name>\fR) associated with each
image in a separate file. This is useful if you want to script
creating HTML pages.
.TP
\fB--html-output\fR <file_name.html>, \fB-l\fR <file_name.html>
Store HTML code that shows HDRHTML viewer for each
image in a separate file. This is useful if you want to script
creating HTML pages.
.SH TEMPLATE FILE FORMAT
pfsouthdrhtml uses two template files \fIhdrhtml_page_templ.html\fR
and \fIhdrhtml_image_templ.html\fR, located in
\fIINSTALL_DIR/share/pfstools/\fR, to generate a web page with an HDR
HTML viewer. The 'page' file contains the HTML of the entire web page
and the 'image' file is used to paste a viewer code for a single
image. You can replace one or both these templates with your own using
\fB--page-template\fR and \fB--image-template\fR options.
.PP
Each template contains HTML code with additional keywords surrounded
by @ marks (@keyword@), which are replaced with HDR HTML specific
code. Most of the keywords are self explanatory, therefore only the
most important are described below.
.TP
@hdr_img_def@ JavaScript objects that must be put in the 'body'
section before any images. These define all the parameters needed to
control HDR HTML viewer.
.TP
@cf_array_def@
Pre-computed array of opacity coefficients. The same array is used for
all images that use the same quality setting. Currently only one such
array could be used per web-page, so images generated with different
quality setting cannot be mixed on a single web page.
.TP
@image_htmlcode@ or @image_htmlcode[base_name]@
Inserts HTML code of all images or a single image with the base_name
(name with no file extension) specified as a parameter. This should be
put where HDR HTML viewer should be located.
.SH EXAMPLES
.TP
pfsin memorial.hdr | pfshdrhtml memorial_church
Generates a web page memorial_church.html with a set of images
memorial_church_*.jpg in the current directory.
.TP
pfsin ~/hdr_images/*.exr | pfssize --maxx 512 --maxy 512 | pfsouthdrhtml hdr_images
Generate a web page with all OpenEXR images from ~/hdr_images/. The
images are resized so that they are not larger than 512x512.
.TP
templ_dir=$INST_DIR/share/pfstools/hdrhtml_hdrlabs_templ/; pfsin img1.hdr img2.exr | pfssize -r 0.2 | pfsouthdrhtml -p ${templ_dir}/hdrhtml_page_templ.html -i ${templ_dir}/hdrhtml_image_templ.html test.html && cp -r ${templ_dir}/hdrhtml_assets ./
The commands above will use an improved template from hdrlabs.com instead of the default one. Note that this template requires html_assets directory to be copied manually to the destination directory. Replace $INST_DIR with the directory where pfstools is installed (/usr/local by default). 
.SH "SEE ALSO"
.BR pfsin (1)
.BR pfsout (1)
.SH BUGS
Please report bugs and comments to the discussion group
http://groups.google.com/group/pfstools


NEWS
This file is not maintained, but required by automake. Visit pfstools
web page to find the latest news:

http://pfstools.sourceforge.net/



src/hdrhtml/pfsouthdrhtml.cpp
/**
 * @brief Create a web page with an HDR viewer
 * 
 * This file is a part of PFSTOOLS package.
 * ---------------------------------------------------------------------- 
 * Copyright (C) 2009 Rafal Mantiuk
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * ---------------------------------------------------------------------- 
 * 
 * @author Rafal Mantiuk, <mantiuk@mpi-sb.mpg.de>
 *
 * $Id: pfsouthdrhtml.cpp,v 1.5 2010/06/13 14:45:55 rafm Exp $
 */

#include "hdrhtml.h"

#include <config.h>

#include <cstdlib>
#include <iostream>
#include <string>

#include <stdio.h>
#include <pfs.h>
#include <getopt.h>


#define PROG_NAME "pfsouthdrhtml"

class QuietException 
{
};

#define min(x,y) ( (x)<(y) ? (x) : (y) )

void printHelp()
{
  fprintf( stderr, PROG_NAME " [<html_page_name>] [--quality <1-5>] [--image-dir <directory_name>] [--page-template <template_file>] [--image-template <template_file>] [--object-output <file_name.js>] [--html-output <file_name.html>] [--verbose] [--help]\n"
    "See the manual page for more information.\n" );
}


void generate_hdrhtml( int argc, char* argv[] )
{
  bool verbose = false;
  int quality = 2;

  // Parse command line parameters
  static struct option cmdLineOptions[] = {
    { "help", no_argument, NULL, 'h' },
    { "verbose", no_argument, NULL, 'v' },
    { "quality", required_argument, NULL, 'q' },
    { "image-dir", required_argument, NULL, 'd' },
    { "object-output", required_argument, NULL, 'o' },
    { "html-output", required_argument, NULL, 'l' },
    { "page-template", required_argument, NULL, 'p' },
    { "image-template", required_argument, NULL, 'i' },
    { NULL, 0, NULL, 0 }
  };
  static const char optstring[] = "q:hvd:o:l:p:i:";
  const char *page_name = NULL;
  const char *image_dir = NULL;
  const char *object_output = NULL;
  const char *html_output = NULL;
  const char *page_template = PKG_DATADIR "/hdrhtml_default_templ/hdrhtml_page_templ.html";
  const char *image_template = PKG_DATADIR "/hdrhtml_default_templ/hdrhtml_image_templ.html";
  
  int optionIndex = 0;
  while( 1 ) {
    int c = getopt_long (argc, argv, optstring, cmdLineOptions, &optionIndex);
    if( c == -1 ) break;
    switch( c ) {
    case 'h':
      printHelp();
      throw QuietException();
    case 'v':
      verbose = true;
      break;
    case 'q':
      quality = strtol( optarg, NULL, 10 );
      if( quality < 1 || quality > 5 )
        throw pfs::Exception( "The quality must be between 1 (worst) and 5 (best)." );
      break;
    case 'd':
      image_dir = optarg;
      break;
    case 'o':
      object_output = optarg;
      break;
    case 'p':
      page_template = optarg;
      break;
    case 'i':
      image_template = optarg;
      break;
    case 'l':
      html_output = optarg;
      break;
    case '?':
      throw QuietException();
    case ':':
      throw QuietException();
    }
  }

  if( optind < argc ) {
    page_name = argv[optind++];    
  }
  if( optind < argc ) {
    throw pfs::Exception( "Too many parameters." );     
  }
  
  pfs::DOMIO pfsio;
 
  bool first_frame = true;

  HDRHTMLSet image_set( page_name, image_dir );
  
  while( true ) {
    pfs::Frame *frame = pfsio.readFrame( stdin );
    if( frame == NULL ) {
      break; // No more frames
    }

    // Get RGB color channels
    pfs::Channel *R= NULL, *B = NULL; // for clarity of the code
    pfs::Channel *Y;
    pfs::Array2DImpl G( frame->getWidth(), frame->getHeight() );
    {
      pfs::Channel *X, *Z;
      
      frame->getXYZChannels( X, Y, Z );
      
      if( X != NULL ) {       // Has color
        R = X;
        B = Z;
        pfs::transformColorSpace( pfs::CS_XYZ, X, Y, Z,
          pfs::CS_RGB, R, &G, B );
      } else {                  // Monochromatic
        Y = frame->getChannel( "Y" );
        if( Y == NULL )
          throw pfs::Exception( "Missing color and luminance information in one of the input frames." );     
        R = Y;
        pfs::copyArray( Y, &G );        
        B = Y;
      }      
    }

    // Get base_name if needed
    std::string base_name;
    {
      const char *filename = frame->getTags()->getString( "FILE_NAME" );
      if( filename == NULL ) {
        if( first_frame && page_name != NULL )
          base_name = page_name;
        else {
          if( first_frame )
            throw pfs::Exception( "Specify page name if FILE_NAME information in missing in the pfsstream" );
          else
            throw pfs::Exception( "Cannot handle multiple images if FILE_NAME information in missing in the pfsstream" );
        }
      } else {
        std::string tmp_str( filename );

        // Remove extension
        int dot_pos = tmp_str.find_last_of( '.' );
        if( dot_pos != std::string::npos & dot_pos > 0 )
          tmp_str = tmp_str.substr( 0, dot_pos );

        // Remove path
        int slash_pos = tmp_str.find_last_of( "/\\" );
        if( slash_pos != std::string::npos )
          tmp_str = tmp_str.substr( slash_pos+1, std::string::npos );
        
        // Substitute invalid characters
        while( true ) {
          int invalid_pos = tmp_str.find_last_of( "-! #@()[]{}`." );
          if( invalid_pos == std::string::npos )
            break;
          tmp_str.replace( invalid_pos, 1, 1, '_' );
        }        
      
        base_name = tmp_str;
      }
      
    }

    std::cerr << "Adding image " << base_name << " to the web page\n";

    image_set.add_image( frame->getWidth(), frame->getHeight(), R->getRawData(), G.getRawData(), B->getRawData(), Y->getRawData(), base_name.c_str(), quality );
    
    pfsio.freeFrame( frame );
    first_frame = false;
  }

  image_set.generate_webpage( page_template, image_template, object_output, html_output );
}


int main( int argc, char* argv[] )
{
  try {
    generate_hdrhtml( argc, argv );
  }
  catch( pfs::Exception ex ) {
    fprintf( stderr, PROG_NAME " error: %s\n", ex.getMessage() );
    return EXIT_FAILURE;
  }
  catch( QuietException  ex ) {
    return EXIT_FAILURE;
  }
  
  return EXIT_SUCCESS;
}
